### PR TITLE
signal handling

### LINKED ISSUE #19: Signal Support
Add Posix signal support as described in our [doc/SIGNALS.md](https://github.com/ComputationalRadiationPhysics/picongpu/blob/0.1.0/doc/SIGNALS.md)

We will probably need a more general **Simulation Control** class to get the communication between the _signal handler_ (thread?), _plugins_ (signal reduce/sync) and the _simulation itself_ clean.

### PR DESCRIPTION
**Pull request by psychocoderHPC**
fix #19

- add new checkpoint parameter `--checkpoint.tryRestart`
- handle incoming signals
- `TimeSlice` extent constructor
- update documentation
- hemera-hzdr: 
  - map Slurm signals to PIConGPU needs
  - send signal 240 seconds before we run into the walltime limit (trigger checkpoint)

The signal mapping in this implementation is different from the suggestion in the issue. 
The reason is that [mpiexec from OpenMPI](https://www.open-mpi.org/doc/v3.0/man1/mpiexec.1.php#toc15) is only propagating a limited number of signals.
Using SIGUSR1 and SIGUSR2 will guarantee the best compatibility with other MPI flavors and different batch systems.

# Review

Best review the last three commits separate.

# Todo

- [x] merge after #3632

### CODE REVIEW

File: include/pmacc/pluginSystem/TimeSlice.hpp [hunk 1]
```diff
@@ -74,6 +74,22 @@ namespace pmacc
...
+
+            /** Constructor
+             *
+             * Select any time step in the range [start,end].
+             *
+             * @param start first time step taken into account
+             * @param end last time step taken into account
+             * @param period select any period(st) time step between start and end
+             */
+            TimeSlice(uint32_t start, uint32_t end = uint32_t(-1), uint32_t period = 1)
```
Comment (sbastrakov) [inline include/pmacc/pluginSystem/TimeSlice.hpp:86]: If `start = 0` by default, the default constructor above seems not required.

File: etc/picongpu/hemera-hzdr/defq.tpl [hunk 2]
```diff
@@ -97,5 +105,5 @@ export OMPI_MCA_io=^ompio
 
 if [ $? -eq 0 ] ; then
   # Run PIConGPU
-  mpiexec --bind-to none !TBG_dstPath/tbg/cpuNumaStarter.sh !TBG_dstPath/input/bin/picongpu !TBG_author !TBG_programParams
+  source !TBG_dstPath/tbg/handleSlurmSignals.sh mpiexec --bind-to none !TBG_dstPath/tbg/cpuNumaStarter.sh !TBG_dstPath/input/bin/picongpu !TBG_author !TBG_programParams
```
Comment (sbastrakov) [inline etc/picongpu/hemera-hzdr/defq.tpl:108]: Is this a single line on purpose?

**Review by sbastrakov — DISMISSED** [review]

File: etc/picongpu/hemera-hzdr/defq.tpl (hunk 2 above)
Comment (psychocoderHPC) [reply etc/picongpu/hemera-hzdr/defq.tpl:108]: I can break with \ at the line end. Will add it tomorrow

Comment (sbastrakov) [reply etc/picongpu/hemera-hzdr/defq.tpl:108]: Maybe it's not required. I was mostly wondering if that was an accidental missing end-of-line. It this is intended, could stay as is

Comment (psychocoderHPC) [reply etc/picongpu/hemera-hzdr/defq.tpl:108]: Wait let me test if I can call it `$(!TBG_dstPath/tbg/handleSlurmSignals.sh mpiexec ...)`
It is required to call it in the same context as the batch script. I tested something else before and accidentally thought I need to use `source`.

File: include/pmacc/pluginSystem/TimeSlice.hpp (hunk 1 above)
Comment (psychocoderHPC) [reply include/pmacc/pluginSystem/TimeSlice.hpp]: updated

File: etc/picongpu/hemera-hzdr/defq.tpl (hunk 2 above)
Comment (psychocoderHPC) [reply etc/picongpu/hemera-hzdr/defq.tpl]: updated

**Comment by psychocoderHPC** [comment]
CI passed, ready to rumble

**Review by sbastrakov — APPROVED** [review]

**Pull request merged**